lib/AST/Expr.cpp

```diff
@@ -1477,6 +1477,20 @@ ArrayExpr *ArrayExpr::create(ASTContext &C, SourceLoc LBracketLoc,
   return new (Mem) ArrayExpr(LBracketLoc, Elements, CommaLocs, RBracketLoc, Ty);
 }
 
+Type ArrayExpr::getElementType() {
+  auto init = getInitializer();
+  if (!init) return Type();
+
+  auto *decl = dyn_cast<AbstractFunctionDecl>(init.getDecl());
+  assert(decl && "Must have AbstractFunctionDecl initializer.");
+
+  return decl->getMethodInterfaceType()
+                         .subst(init.getSubstitutions())
+                         ->getAs<AnyFunctionType>()
+                         ->getParams()[0]
+                         .getPlainType();
```

Comment:
It would be cheapest to substitute the smallest structural component, ie, put the subst() call on the result of getPlainType().